lib/blob/blobstore.c

```diff
@@ -5985,6 +5985,28 @@ spdk_blob_reset_used_clusters_cache(struct spdk_blob *blob)
 		spdk_spin_unlock(&blob->bs->used_lock);
 	}
 }
+
+struct spdk_bit_array *
+spdk_blob_get_cluster_bitmap(struct spdk_blob *blob)
```

Comment:
Shouldn't we be doing this by holding spinlock blob->bs->used_lock, preferably inside here instead of caller. The active clusters cache is a dirty list which could be continuously getting modified.

Reply:
snapshots are readonly so I thought this wouldn't be required, but actually this function is not specifically for blob, so we should take the lock, I'll modify it.

Reply:
Done, PTAL

Reply:
One concern is that bit array creation looks to allocate memory underneath. And allocating memory under spinlock is generally forbidden. It'd be a good idea to do bitarray creation outside lock using the `clean.num_clusters` or `active.num_clusters` which should ideally be in sync(and otherwise static, unless blob has been resized meanwhile).
cc: @dsavitskiy

Reply:
Yeah, good point..
But I'm also wondering, is this lock the correct one?
Perhaps we should freeze the blob first, then get the clusters?

Reply:
After discussion with the folks we decided to add RO support for now, with freeze on another ticket.
I've added callback based approach which will be required for freeze, PTAL